File: packages/python-sdk/hyrelog/client/company.py
"""
Company-level client for read-only operations
"""

from typing import Optional
from opentelemetry import trace

from hyrelog.client.base import BaseClient
from hyrelog.types import QueryOptions, QueryResponse, HyreLogClientOptions


class HyreLogCompanyClient(BaseClient):
    """Company client for querying events across workspaces"""

    def __init__(
        self,
        company_key: str,
        base_url: str = "https://api.hyrelog.com",
        debug: bool = False,
        timeout: float = 30.0,
        retry_config: Optional[dict] = None,
    ):
        options = HyreLogClientOptions(
            api_key=company_key,
            base_url=base_url,
            debug=debug,
            timeout=timeout,
            retry_config=retry_config,
        )
        super().__init__(options)

    async def query_events(self, options: Optional[QueryOptions] = None) -> QueryResponse:
        """Query events across all workspaces in the company"""
        if options is None:
            options = QueryOptions()

        tracer = trace.get_tracer("hyrelog-sdk")
        span = tracer.start_span("hyrelog.query_company_events")

        try:
            params = options.model_dump(exclude_none=True, by_alias=True)
            result = await self._request("GET", "/v1/key/company/events", params=params)

            span.set_status(trace.Status(trace.StatusCode.OK))
            return QueryResponse(**result)
        except Exception as e:
            span.record_exception(e)
            span.set_status(trace.Status(trace.StatusCode.ERROR, str(e)))
            raise

    async def query_global_events(self, options: Optional[QueryOptions] = None) -> QueryResponse:
        """Query events globally across all regions (Phase 3 feature)"""
        if options is None:
            options = QueryOptions()

        tracer = trace.get_tracer("hyrelog-sdk")
        span = tracer.start_span("hyrelog.query_global_events")

        try:
            params = options.model_dump(exclude_none=True, by_alias=True)
            result = await self._request("GET", "/v1/key/company/events/global", params=params)

            span.set_status(trace.Status(trace.StatusCode.OK))
            return QueryResponse(**result)
        except Exception as e:
            span.record_exception(e)
            span.set_status(trace.Status(trace.StatusCode.ERROR, str(e)))
            raise

    async def get_regions(self) -> dict:
        """Get company region information"""
        tracer = trace.get_tracer("hyrelog-sdk")
        span = tracer.start_span("hyrelog.get_regions")

        try:
            result = await self._request("GET", "/v1/key/company/regions")

            span.set_status(trace.Status(trace.StatusCode.OK))
            return result
        except Exception as e:
            span.record_exception(e)
            span.set_status(trace.Status(trace.StatusCode.ERROR, str(e)))
            raise

File: packages/node-sdk/src/client/company.ts
/**
 * Company-level client for read-only operations
 */

import { BaseClient } from "./base";
import type { Event, QueryOptions, QueryResponse, HyreLogClientOptions } from "../types";
import { trace } from "@opentelemetry/api";

/**
 * Company client for querying events across workspaces
 */
export class HyreLogCompanyClient extends BaseClient {
  constructor(options: { companyKey: string } & Omit<HyreLogClientOptions, "apiKey">) {
    super({ ...options, apiKey: options.companyKey });
  }

  /**
   * Query events across all workspaces in the company
   */
  async queryEvents(options: QueryOptions = {}): Promise<QueryResponse> {
    const tracer = trace.getTracer("hyrelog-sdk");
    const span = tracer.startSpan("hyrelog.queryCompanyEvents");

    try {
      const params = new URLSearchParams();
      if (options.page) params.set("page", options.page.toString());
      if (options.limit) params.set("limit", options.limit.toString());
      if (options.from) params.set("from", options.from);
      if (options.to) params.set("to", options.to);
      if (options.action) params.set("action", options.action);
      if (options.category) params.set("category", options.category);
      if (options.actorId) params.set("actorId", options.actorId);
      if (options.actorEmail) params.set("actorEmail", options.actorEmail);
      if (options.workspaceId) params.set("workspaceId", options.workspaceId);
      if (options.projectId) params.set("projectId", options.projectId);

      const queryString = params.toString();
      const path = `/v1/key/company/events${queryString ? `?${queryString}` : ""}`;

      const result = await this.request<QueryResponse>("GET", path);

      span.setStatus({ code: 1 });
      span.end();
      return result;
    } catch (error) {
      span.recordException(error instanceof Error ? error : new Error(String(error)));
      span.setStatus({ code: 2, message: error instanceof Error ? error.message : String(error) });
      span.end();
      throw error;
    }
  }

  /**
   * Query events globally across all regions (Phase 3 feature)
   */
  async queryGlobalEvents(options: QueryOptions = {}): Promise<QueryResponse> {
    const tracer = trace.getTracer("hyrelog-sdk");
    const span = tracer.startSpan("hyrelog.queryGlobalEvents");

    try {
      const params = new URLSearchParams();
      if (options.page) params.set("page", options.page.toString());
      if (options.limit) params.set("limit", options.limit.toString());
      if (options.from) params.set("from", options.from);
      if (options.to) params.set("to", options.to);
      if (options.action) params.set("action", options.action);
      if (options.category) params.set("category", options.category);
      if (options.actorId) params.set("actorId", options.actorId);
      if (options.actorEmail) params.set("actorEmail", options.actorEmail);
      if (options.workspaceId) params.set("workspaceId", options.workspaceId);
      if (options.projectId) params.set("projectId", options.projectId);

      const queryString = params.toString();
      const path = `/v1/key/company/events/global${queryString ? `?${queryString}` : ""}`;

      const result = await this.request<QueryResponse>("GET", path);

      span.setStatus({ code: 1 });
      span.end();
      return result;
    } catch (error) {
      span.recordException(error instanceof Error ? error : new Error(String(error)));
      span.setStatus({ code: 2, message: error instanceof Error ? error.message : String(error) });
      span.end();
      throw error;
    }
  }

  /**
   * Get company region information
   */
  async getRegions(): Promise<{
    company: {
      id: string;
      name: string;
      primaryRegion: string;
      replicaRegions: string[];
    };
    regions: Array<{
      region: string;
      isPrimary: boolean;
      isReplica: boolean;
      dbUrl: string;
      readOnlyUrl: string;
      coldStorageProvider: string;
      coldStorageBucket: string;
      health: string;
    }>;
  }> {
    const tracer = trace.getTracer("hyrelog-sdk");
    const span = tracer.startSpan("hyrelog.getRegions");

    try {
      const result = await this.request<{
        company: {
          id: string;
          name: string;
          primaryRegion: string;
          replicaRegions: string[];
        };
        regions: Array<{
          region: string;
          isPrimary: boolean;
          isReplica: boolean;
          dbUrl: string;
          readOnlyUrl: string;
          coldStorageProvider: string;
          coldStorageBucket: string;
          health: string;
        }>;
      }>("GET", "/v1/key/company/regions");

      span.setStatus({ code: 1 });
      span.end();
      return result;
    } catch (error) {
      span.recordException(error instanceof Error ? error : new Error(String(error)));
      span.setStatus({ code: 2, message: error instanceof Error ? error.message : String(error) });
      span.end();
      throw error;
    }
  }
}

